### PR TITLE
feat: Reduce memory consumption by storing array length as `u32` during SSA

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

We're currently representing the length of a `Type::Array` using a usize. This means that despite the fact that Noir indexes its arrays with `u32`s we need to use 64bits to store this length within rust (assuming the user is on a 64bit system).

This PR switches how we store `Type`s within SSA so that we always store an array length as a u32 which saves significant amounts of memory (100 MB reduction on compiling `eddsa`).

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/ir/dfg.rs [hunk 1]
```diff
@@ -307,13 +307,20 @@ impl DataFlowGraph {
...
+            .map(|(position, typ)| {
+                self.values.insert(Value::Instruction {
+                    typ,
+                    position,
+                    instruction: instruction_id,
+                })
+            })
+            .collect();
+
+        self.results.insert(instruction_id, results);
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/ir/dfg.rs:323]: Also includes this change which inlines an unnecessary function and avoid repeatedly pushing onto a `Vec` in favour of iterators.

**Comment by jfecher** [comment]
> This PR switches how we store Types within SSA so that we always store an array length as a u32 which saves significant amounts of memory (100 MB reduction on compiling eddsa).

That must be a lot of array types being created for a 32 byte change to have such an impact. Maybe we should consider caching array types (or nested types in general) if they have such an impact.

File: compiler/noirc_evaluator/src/ssa/ir/dfg.rs [hunk 2]
```diff
@@ -307,13 +307,20 @@ impl DataFlowGraph {
...
+            .into_iter()
+            .enumerate()
+            .map(|(position, typ)| {
+                self.values.insert(Value::Instruction {
+                    typ,
+                    position,
+                    instruction: instruction_id,
+                })
+            })
+            .collect();
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/ir/dfg.rs:321]:
```suggestion
        let result_types = self.instruction_result_types(instruction_id, ctrl_typevars);
        let results = vecmap(result_types.into_iter().enumerate(), |(position, typ)| {
            let instruction = instruction_id; // could also rename the parameter
            self.values.insert(Value::Instruction { typ, position, instruction })
        });
```

**Review by jfecher — APPROVED** [review]

**Comment by TomAFrench** [comment]
On second thoughts, I think we should hold off on merging this PR (and similar memory-motivated PRs) until we have #6604 .

**Pull request merged**